### DISCUSSION TITLE
Consider removing support of "dims on the fly"

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-devs/pymc3#6065] [General]
Right now we can specify dims on the fly, which are taken from the shape of a variable:

```python
import pymc as pm
with pm.Model() as m:
  x = pm.Normal("x", shape=(3,), dims=("trial",))
  print(m.dim_lengths["trial"].eval())  # 3
```

This is neat, but causes some code complexity, because we also allow dims to inform the shape of a variable.
```python
with pm.Model(coords={"trial": range(3)}) as m:
  x = pm.Normal("x", dims=("trial",))
  print(x.eval().shape)  # (3,)
```

Which forces us to create new variables in a two-step process, first without dim information and then with it (via resizing). Code could be simplified if we don't have to worry about dims being specified on the fly. 

These two behaviors of dims are also potentially confusing when mixed together:
```python
with pm.Model() as m:
  x = pm.Normal("x", shape=5, dims=("trial",))
  y = pm.Normal("y", dims=("trial",))
  print(y.eval().shape)  # (5,)
```

**Comment by aseyboldt** [comment]
I didn't realize this would work :-)
I think it makes sense to have to define the dim (with coords) before you use it (either in pm.Model`, through pm.Data or methods on the model). "Explicit is better than implicit" :-)

Do you have an idea if that is used a lot by users at the moment however? Since when is this possible (it wasn't possible from the start)?

**Reply by ricardoV94** [reply]
This is a V4 thing. I don't think it was advertised that much, so hopefully adoption is equally low.

**Comment by ferrine** [comment]
I didn't know it was a thing

**Comment by twiecki** [comment]
Seems pretty clear we're fine to remove it then.

**Comment by michaelosthege** [comment]
This was done in #6112 and released in v4.2.0.
